server/datastore/mysql/migrations/tables/20201011162341_CleanupSoftDeletedColumns.go

```diff
@@ -0,0 +1,95 @@
+package tables
+
+import (
+	"fmt"
+
+	"database/sql"
+	"github.com/kolide/goose"
+)
+
+func init() {
+	goose.AddMigration(Up20201011162341, Down20201011162341)
+}
+
+func cleanupSoftDeleteFields(tx *sql.Tx, dbTable string) error {
+	deleteStmt := fmt.Sprintf("DELETE FROM `%s` WHERE deleted;", dbTable)
+
+	_, err := tx.Exec(deleteStmt)
+	if err != nil {
+		return err
+	}
+
+	for _, column := range []string{"deleted", "deleted_at"} {
+		alterStmt := fmt.Sprint(
+			"ALTER TABLE `%s` DROP COLUMN `%s`;", dbTable, column)
+
+		_, err := tx.Exec(alterStmt)
+		if err != nil {
+			return err
+		}
+	}
+
+	return nil
+}
+
+func addSoftDeleteFields(tx *sql.Tx, dbTable string) error {
+	addDeletedStmt := fmt.Sprint(
+		"ALTER TABLE `%s` "+
+			"ADD COLUMN `deleted` TINYINT(1) NOT NULL DEFAULT FALSE;",
+		dbTable)
+	_, err := tx.Exec(addDeletedStmt)
+	if err != nil {
+		return err
+	}
+
+	addDeletedAtStmt := fmt.Sprint(
+		"ALTER TABLE `%s` "+
+			"ADD COLUMN `deleted_at` TIMESTAMP NULL DEFAULT NULL;",
+		dbTable)
+	_, err = tx.Exec(addDeletedAtStmt)
+	if err != nil {
+		return err
+	}
+
+	return nil
+}
+
+func getTablesForCleanupSoftDeletedColumnsMigration() []string {
+	tables := []string{
+		"distributed_query_campaigns",
+		"labels",
+		"invites",
+		"hosts",
+		"packs",
+		"queries",
+		"scheduled_queries",
+		"users",
+	}
+	return tables
+}
+
+func Up20201011162341(tx *sql.Tx) error {
+	tables := getTablesForCleanupSoftDeletedColumnsMigration()
+
+	for _, table := range tables {
+		err := cleanupSoftDeleteFields(tx, table)
+		if err != nil {
+			return err
+		}
+	}
+
+	return nil
+}
+
+func Down20201011162341(tx *sql.Tx) error {
```

Comment:
Fleet doesn't actually have a mechanism for running down-migrations, and most migrations don't actually implement the down portion. Feel free to skip it next time. Do we need to document this?

Reply:
Ah, yeah, I guess I didn't think about migrating to old versions too much. However, it's interesting that you say most migrations don't implement this. I was looking at the existing migrations and it actually looks like they usually do (or at least the ones I've checked).

Like, if the Up migration adds a column, the Down one will drop the column. Or if Up Creates a table, then Down will drop the table.

I think it would be good to document this either way, but also consider supporting running down-migrations.

In any case - I imagine that's probably out of scope for this PR. I'd just as soon keep the Down migration code in unless we're going to be dropping all of the Down migration functions.

Reply:
You are right that for some time we were writing down-migrations. Eventually we stopped when we realized that we had gone years without needing them or even implementing a way to run them.

I think it's totally fine to include down-migrations in the code if only as documentation. Maybe someday we will remove them.

One of the big problems with actually running down-migrations is that the migrations tend to be lossy and running down-migrations is likely not to return things to the same state. In the unlikely event that the version needs to be downgraded I think it's best for users to restore from a DB backup or manually determine the appropriate commands to run to return the schema to a consistent state.